### PR TITLE
loader: support schema/table sharding

### PR DESCRIPTION
**Pull request by holys**
@iamxy  @GregoryIan  PTAL

### CODE REVIEW

File: loader/config.go [hunk 1]
```diff
@@ -92,6 +92,16 @@ type Config struct {
 
 	configFile          string
 	SkipConstraintCheck int `toml:"skip-unique-check" json:"skip-unique-check"`
+
+	RouteRules []*RouteRule `toml:"route-rules" json:"route-rules"`
+}
+
+// RouteRule is the route rule for loading schema and table into specified schema and table.
+type RouteRule struct {
+	SchemaPattern string `toml:"schema-pattern" json:"schema-pattern"`
```
Comment (GregoryIan) [inline loader/config.go]: `PatternSchema`?

File: loader/loader.go [hunk 2]
```diff
@@ -262,9 +284,19 @@ func (l *Loader) Restore() error {
 	return nil
 }
 
+func genRouter(rules []*RouteRule) (tableRouter route.TableRouter) {
+	tableRouter = route.NewTrieRouter()
+
+	for _, rule := range rules {
+		tableRouter.Insert(rule.SchemaPattern, rule.TablePattern, rule.TargetSchema, rule.TargetTable)
```
Comment (GregoryIan) [inline loader/loader.go]: need handle error

File: loader/loader.go [hunk 3]
```diff
@@ -24,13 +24,21 @@ import (
...
 	"github.com/ngaut/log"
+	"github.com/pingcap/tidb-tools/pkg/tableroute"
 )
 
 var (
 	jobCount      = 1000
 	maxRetryCount = 10
 )
 
+type restoreType byte
```
Comment (GregoryIan) [inline loader/loader.go]: seem useless

File: loader/loader.go [hunk 4]
```diff
@@ -424,8 +458,20 @@ func (l *Loader) restoreSchema(conn *Conn, sqlFile string, schema string) error
 					continue
 				}
 
+				// rename schema
+				if typ == restoreSchema {
```
Comment (GregoryIan) [inline loader/loader.go]: typ seed useless, remove it. just leave `query = renameShardingSchema(l.tableRouter, query, schema, table)`

File: loader/loader.go [hunk 5]
```diff
@@ -441,6 +487,35 @@ func (l *Loader) restoreSchema(conn *Conn, sqlFile string, schema string) error
 	return nil
 }
 
+// renameShardingTable renames table name like `table-001`` to `table` by modifying query.
+func renameShardingTable(router route.TableRouter, query, schema, table string) string {
```
Comment (GregoryIan) [inline loader/loader.go]: ditto

File: loader/loader.go [hunk 6]
```diff
@@ -441,6 +487,35 @@ func (l *Loader) restoreSchema(conn *Conn, sqlFile string, schema string) error
...
+
+func fetchMatchedLiteral(router route.TableRouter, schema, table string) (string, string) {
+	if schema == "" {
+		return schema, table
+	}
+	targetSchema, targetTable := router.Match(schema, table)
+	if targetSchema == "" {
+		return schema, table
+	}
+
```
Comment (GregoryIan) [inline loader/loader.go]: we have rule
```
pattern-schema = "route_*"
pattern-table = "abc_*"
target-schema = "route"
target-table = "abc"
```
if I call `fetchMatchedLiteral(router, "route_1", "xyz")`, what should it return?

**Comment by GregoryIan** [comment]
run `make check` , and format the code

File: loader/loader.go [hunk 7]
```diff
@@ -441,6 +487,35 @@ func (l *Loader) restoreSchema(conn *Conn, sqlFile string, schema string) error
...
+func renameShardingSchema(router route.TableRouter, query, schema, table string) string {
+	targetSchema, targetTable := fetchMatchedLiteral(router, schema, table)
+	log.Debugf("table router match. origin_table:%s,targetTable:%s, origin_schema:%s, targetSchema:%s,query:%-.100v",
+		table, targetTable, schema, targetSchema, query)
+
+	return SQLReplace(query, schema, targetSchema)
+}
+
+func fetchMatchedLiteral(router route.TableRouter, schema, table string) (string, string) {
```
Comment (GregoryIan) [inline loader/loader.go]: handle schema/table case sensitive

Comment (jomenxiao) [inline loader/loader.go]: please  let output same with config.toml

Comment (holys) [inline loader/loader.go]: @jomenxiao  for tidb, schema/table is case insensitive. 

Comment (GregoryIan) [inline loader/loader.go]: keep the origin schema/table name. If we need to load data from tidb to some heterogeneous databases, It could cause error.

Comment (holys) [inline loader/loader.go]: @jomenxiao  comment addressed

File: loader/loader.go [hunk 8]
```diff
@@ -428,10 +460,19 @@ func (l *Loader) restoreSchema(conn *Conn, sqlFile string, schema string) error
 				}
 
 				var sqls []string
-				if len(schema) > 0 {
+
+				if table != "" {
+					// for table
+					query = renameShardingTable(l.tableRouter, query, schema, table)
```
Comment (GregoryIan) [inline loader/loader.go]: `fetchMatchedLiteral` always follows `renameShardingTable`, and we call `fetchMatchedLiteral ` in `renameShardingTable`.

**Comment by jomenxiao** [comment]
LGTM

**Comment by GregoryIan** [comment]
Rest LGTM

**Pull request merged**